docs: add "deploy application using an incoming hook" post (#674)
devopness/devopness: docs/applications/deploy-application-using-incoming-hook.md

---

> title: Deploy Application using an Incoming Hook

Reviewer: The same way we separated SSL, DNS and `find server IP address`, I believe here we should first document an independent topic guiding users on how to create incoming webhooks (`/docs/webhooks/create-incoming-webhook.md`) and then here go through the steps that are specific to application deployments, with link to the common `create-incoming-webhook` steps.

Suggestion of intro text for `create-incoming-webhook`:
```
title: Create Incoming Webhooks
intro: Incoming Webhooks are a simple way to integrate external services with your Devopness environments, 
using a unique URL through which a JSON payload can be sent to run a pipeline with custom input data. 
Some use cases of external services events include "deploy applications automatically based on 
GitHub/GitLab/Bitbucket repository updates", "provision and scale infrastructure resources when you 
payment gateway confirms that a custom subscription has been processed", and many more use cases 
that can benefit from webhooks integration. Learn how to reate an incoming webhook to run pipelines 
programmatically when events happen on external services.
```

Author: @jfoliveira PR updated

links:
    overview:
    quickstart:
    previous:
    next:
    guides:
    related:
    featured:
---

Devopness is an API-First product, meaning that every feature in our product is first designed with the API usage in mind. That development approach implies that every action you make using our web app can also be directly made using API calls.

Webhooks, for now, are an API only feature; so this post will guide you through the usage of our API to help you create a webhook.

Before creating our incoming webhook, we need the `Application ID` (`<application_id>`) and `Pipeline ID` (`<pipeline_id>`) to use as parameters in our API calls.

<details open>
  <summary>Steps to find `Pipeline ID` and `Application ID`</summary>

1. On Devopness, navigate to a project then select an environment
1. Find the `Applications` card
1. Click `View` in the `Applications` card to see a list of existing `Applications`
1. In the list of applications find the application with the pipeline you want to manage and click `DETAILS`
1. On the upper-right corner of the applications details view click `SETTINGS`
1. Use the drop-down menu to choose `PIPELINES`

> 1. In the list of pipelines find the pipeline you want to manage and click `DETAILS`

Author: ```suggestion
1. In the list of pipelines, find the pipeline you want to manage and click `DETAILS`
```

    > If you haven't created a pipeline yet, follow the [Add a Pipeline](/docs/pipelines/add-pipeline) guide
1. Copy the `Application ID` and `Pipeline ID` from the URL, considering the following URL format:
```bash
https://app.devopness.com/projects/<project_id>/environments/<environment_id>/applications/<application_id>/pipelines/<pipeline_id>
```

</details>

Once you have your `Application ID` (`<application_id>`) and `Pipeline ID` (`<pipeline_id>`), please follow the instructions below to add a incoming webhook to your pipeline:

1. In a terminal window, run [POST /users/login](https://api-docs.devopness.com/#tag/Users/operation/loginUser) request using your Devopness account email and password
    ```bash
    curl --request POST \
      --url https://api.devopness.com/users/login \
      --header 'Accept: application/json' \
      --header 'Content-Type: application/json' \
      --data '{
    	"email": "<email>",
    	"password": "<password>"
    }'
    ```
1. From the previous command response, copy the field `access_token`

> 1. In a terminal window, run [POST /hooks/incoming](https://api-docs.devopness.com/#tag/Hooks/operation/addHook) request, replacing `<application_id>` and `<pipeline_id>`.

Author: ```suggestion
1. In a terminal window, run [POST /hooks/incoming](https://api-docs.devopness.com/#tag/Hooks/operation/addHook) request, replacing `<access_token>`, `<application_id>` and `<pipeline_id>`.
```

  <summary>Steps to find `Application ID` and `Deploy Pipeline ID`</summary>

1. On Devopness, navigate to a project then select an environment
1. Find the `Applications` card
1. Click `View` in the `Applications` card to see a list of existing `Applications`
1. In the list of applications, find the application with the pipeline you want to manage and click `DETAILS`
1. On the upper-right corner of the applications details view click `SETTINGS`
1. Use the drop-down menu to choose `PIPELINES`
1. In the list of pipelines, find the pipeline you want to manage and click `DETAILS`
    > If you haven't created a pipeline yet, follow the [Add a Pipeline](/docs/pipelines/add-pipeline) guide
1. Copy the `Application ID` and `Deploy Pipeline ID` from the URL, considering the following URL format:
    ```bash
    https://app.devopness.com/projects/<project_id>/environments/<environment_id>/applications/<application_id>/pipelines/<pipeline_id>
    ```

</details>

Once you have your `Application ID` (`<application_id>`) and `Deploy Pipeline ID` (`<pipeline_id>`), please follow the instructions below to add an incoming webhook to your application that triggers the deploy pipeline:

> 1. Follow the steps on [Create an Incoming Webhook](/docs/webhooks/create-incoming-webhook) guide 
> 2. Once your web hook is created, copy the webhook the hook's unique URL (`url` field)
> 3. Take note of the hook secret signature key (`secret` field)
>   > If the hook was created without a `secret` (non recommended), you can skip this step

Author: ```suggestion
1. Follow the steps on [Create an Incoming Webhook](/docs/webhooks/create-incoming-webhook) guide 
1. Once your web hook is created, copy the webhook the hook's unique URL (`url` field)
1. Take note of the hook secret signature key (`secret` field)
  > If the hook was created without a `secret` (non recommended), you can skip this step
```

1. Add your webhook to your source provider

>     - Follow Bitbucket' [Manage webhooks: create webhooks](https://support.atlassian.com/bitbucket-cloud/docs/manage-webhooks/#Create-webhooks) guide; or

Reviewer: I believe the links should be easier to maintain and the usage of `guide, or` in the end of each item of the bullet list will make the lines harder to edit.

Suggestions below:

~~1. Add your webhook to your source provider~~
1. Add your webhook to the source provider where each application's source code is hosted, by following the source provider's instructions on the links below:

```Suggestion
- Bitbucket: [Manage webhooks: create webhooks](https://support.atlassian.com/bitbucket-cloud/docs/manage-webhooks/#Create-webhooks)

- Github: [Webhooks: setting up a webhook](https://docs.github.com/en/webhooks-and-events/webhooks/creating-webhooks#setting-up-a-webhook)

- Gitlab: [Webhooks: configure a webhook in GitLab](https://docs.gitlab.com/ee/user/project/integrations/webhooks.html#configure-a-webhook-in-gitlab)
```

Author: Suggestion Applied

1. Find the `Applications` card
1. Click `View` in the `Applications` card to see a list of existing `Applications`
1. In the list of applications, find the application with the pipeline you want to manage and click `DETAILS`
1. On the upper-right corner of the applications details view click `SETTINGS`
1. Use the drop-down menu to choose `PIPELINES`
1. In the list of pipelines, find the pipeline you want to manage and click `DETAILS`
    > If you haven't created a pipeline yet, follow the [Add a Pipeline](/docs/pipelines/add-pipeline) guide
1. Copy the `Application ID` and `Deploy Pipeline ID` from the URL, considering the following URL format:
    ```bash
    https://app.devopness.com/projects/<project_id>/environments/<environment_id>/applications/<application_id>/pipelines/<pipeline_id>
    ```

</details>

Once you have your `Application ID` (`<application_id>`) and `Deploy Pipeline ID` (`<pipeline_id>`), please follow the instructions below to add an incoming webhook to your application that triggers the deploy pipeline:

1. Create an incoming webhook using the payload below, replacing `<application_id>` and `<pipeline_id>`

>     > Follow the [Create an Incoming Webhook](/docs/webhooks/create-incoming-webhook) guide

Reviewer: I stick to my point in previous comments. Given that:
1. The reader already obtained `<application_id>` and `<pipeline_id>`
2. This topic is already pointing the reader to the `create-incoming-webhook` topic
3. The payload is being exemplified and explained there

Why do we need to repeat the payload example here?

Author: @jfoliveira currently both posts are using similar but different payloads, being the one on the create incoming hook simpler; I could replace the payload there using this one.

I believe this would solve the matter being discussed here.

Will update the PR soon.

1. Use the drop-down menu to choose `PIPELINES`
1. In the list of pipelines, find the pipeline you want to manage and click `DETAILS`
    > If you haven't created a pipeline yet, follow the [Add a Pipeline](/docs/pipelines/add-pipeline) guide
1. Copy the `Application ID` and `Deploy Pipeline ID` from the URL, considering the following URL format:
    ```bash
    https://app.devopness.com/projects/<project_id>/environments/<environment_id>/applications/<application_id>/pipelines/<pipeline_id>
    ```

</details>

Once you have your `Application ID` (`<application_id>`) and `Deploy Pipeline ID` (`<pipeline_id>`), please follow the instructions below to add an incoming webhook to your application that triggers the deploy pipeline:

1. In a terminal window, submit a request to Devopness API endpoint `POST /hooks/incoming`, replacing `<access_token>`, `<application_id>` and `<pipeline_id>`.
    > If this is your first time calling the Devopness API, follow the [Create an Incoming Webhook](/docs/webhooks/create-incoming-webhook) guide to see instructions on getting your `<access_token>`.
    ```bash
    curl --request POST \
      --url https://api.devopness.com/hooks/incoming \

>       --header 'Accept: application/json' \

Reviewer: If the reader will follow the `create-incoming-webhook` guide, the `POST /hooks/incoming` HTTP request is already documented in that guide.
Given that, do we still need to repeat this HTTP request here?

Reviewer: See previous comment that suggested that: https://github.com/devopness/devopness/pull/674/files#r1204504083

Once you have your `Application ID` (`<application_id>`) and `Deploy Pipeline ID` (`<pipeline_id>`), please follow the instructions below to add an incoming webhook to your application that triggers the deploy pipeline:

> 1. Follow the [Create an Incoming Webhook](/docs/webhooks/create-incoming-webhook) guide and copy the fields `url` (unique hook URL) and `secret` (signature key)

Reviewer: Suggesting `take note` instead of copy, as copy a second value would replace a previous value copied to the clipboard.

```suggestion
1. Follow the steps on [Create an Incoming Webhook](/docs/webhooks/create-incoming-webhook) guide 
2. Once your web hook is created, copy the webhook the hook's unique URL (`url` field)
3. Take note of the hook secret signature key (`secret` field)
  > If the hook was created without a `secret` (non recommended), you can skip this step
```
